base/compiler/typeinfer.jl

```diff
@@ -321,7 +321,7 @@ function CodeInstance(interp::AbstractInterpreter, result::InferenceResult,
             t = @_gc_preserve_begin inferred_result
             relocatability = unsafe_load(unsafe_convert(Ptr{UInt8}, inferred_result), Core.sizeof(inferred_result))
             @_gc_preserve_end t
-        elseif inferred_result === nothing
+        elseif !(inferred_result isa CodeInfo)
```

Comment:
This is the actual fix. The other changes are the refinements on the test cases (which were done in #53478).

Reply:
Hm but on master we still have https://github.com/JuliaLang/julia/blob/b3b27360672f87af78e90f43737e32988d9b8be1/base/compiler/typeinfer.jl#L331

So I am not sure how it would be fixed there. IIRC even on master I needed a abs int overload that set relocatability?

Reply:
The relocatability indicates that we know the `inferred` field is safe to duplicate, which doesn't seem would be the case if this is not one of the expected objects? (i.e. String or Nothing)